### ISSUE TITLE
Restore is slow

### ISSUE THREAD

**Issue by sirupsen** [benbjohnson/litestream#266]
Hey @benbjohnson, long time no see!! Thank you for working on Litestream! 🙏🏻 

I, too, love Sqlite. I wanted to track a few events on my website, e.g. what people search for, and saw this as an opportunity to use Litestream. Loved the idea of tracking events in Sqlite and just do analysis on a local copy.

However, even though my db is only ~100kb on disk and ~1000 rows over a few days, it takes ~10 seconds to restore with `litestream restore`, and this is fastly going up. 

Is there a plan for a `litestream compress` or similar to avoid replaying the WAL from early on? Or am I doing something wrong? Unfortunately this will be a bit of a deal-breaker to me using this in production :(

**Comment by benbjohnson** [comment]
hey @sirupsen! 👋 It's been a long time indeed! Still at Shopify?

Restore performance is something that needs improvement but I'm surprised it's 10s for such a small workload. You can change the `retention` and `snapshot-interval` configuration fields to create a snapshot more frequently. If you don't care about keeping historic data, you can just set `retention` to `1h` and it'll just keep one snapshot that's an hour old at most.

If you do want to retain data longer, then you could set `retention` to `24h` (which is the default anyway) and the `snapshot-interval` to `1h`. That'll make a new snapshot every hour but only keep them for a rolling 24 hours.

Another option you could try is setting the `-parallel N` flag on the `litestream restore` command. If you set that to something high like `64` then it should speed up the downloads at least.

Finally, there are plans for maintaining a hot backup so you can restore instantly but that's still a few months away. I'm also working on a version that works in a serverless environment but that's probably going to be ready later in the year.

**Comment by sirupsen** [comment]
I stopped working at Shopify mid last year, doing infra consulting now :)

Thank youuuu! `snapshot-interval` is exactly what I needed... I somehow missed that in the docs. I can add a section to the docs in the `Tips & Caveats` section if you'd accept it?

FWIW I'm using Cloud Run, so for me, it's already working in serverless ;)

**Comment by benbjohnson** [comment]
> _I can add a section to the docs in the Tips & Caveats section if you'd accept it?_

Yes! That'd be awesome. Thanks, Simon.

> _FWIW I'm using Cloud Run, so for me, it's already working in serverless_

Cool. I saw some folks talking about getting Litestream running on Cloud Run but I haven't had a chance to give it a go yet. 

The idea of "serverless SQLite" that I'm thinking of is paging in data on-demand in a way that's transactionally safe. That way it'd give you zero startup time but also low-latency queries once data is hot on a serverless instance. I'm still toying around with the idea but I think it might have some legs.

**Comment by benbjohnson** [comment]
@sirupsen I saw your comment in https://github.com/benbjohnson/litestream/discussions/223#discussioncomment-1977888 but I'm moving the discussion back over to this ticket.

> _I am seeing it being stuck on restore too. Is there a good debugging step I can take? 👀_

Can you hit `CTRL-\` to issue a `SIGQUIT` when it gets stuck for a bit? That should dump out a stack trace that'll tell us what it's stuck on.

**Comment by sirupsen** [comment]
Yeah, for sure, this is the [stacktrace I get](https://gist.github.com/sirupsen/655fafdc35d86cb7dd2860e7c645cc81) from sending `SIGQUIT` just before it exits. It exits after just a second or two (with that error), it doesn't get stuck in a loop 👀  I got a few stacktraces though sending `SIGQUIT` before it terminates, and they all look like that

I have nothing sensitive in this database, so if this isn't enough for you to figure it out, ping me on your Slack and I'll give you access to this bucket 👍🏻

**Comment by benbjohnson** [comment]
@sirupsen The `missing initial wal segment` issue from @pfw turned out to be two applications replicating into the same bucket and their retention enforcement was deleting each other's WAL segments: https://github.com/benbjohnson/litestream/issues/224#issuecomment-881794809

I think the issue might be that GCR doesn't enforce a single instance at a time and there could be overlap—especially when deploying—that's causing issues. I think GCR isn't going to work well until I can get better support for serverless in Litestream.

I'm not sure if you're committed to GCR but another good alternative is [fly.io](https://fly.io). If you attach a persistent disk on their instances then they enforce a single instance at a time.

**Comment by sirupsen** [comment]
Fair enough... I will consider migrating to fly.io.

How do I fix this error though even when nothing is running on GCP? To recover my dear database

**Comment by benbjohnson** [comment]
> _How do I fix this error though even when nothing is running on GCP? To recover my dear database_

Unfortunately, with the missing initial WAL segment the best you can do is recover from the last snapshot.

```sh
# Copy out the last snapshot.
cp generations/f6d6d1e96d38dafb/snapshots/00000093.snapshot.lz4 db.lz4

# Uncompress it
lz4 db.lz4

# Verify the database integrity
sqlite3 db
sqlite> PRAGMA integrity_check;
ok
```

**Comment by sirupsen** [comment]
I don't deserve you, thank you :)

https://github.com/benbjohnson/litestream.io/pull/43

**Comment by benbjohnson** [comment]
Thanks for going on this debugging journey with me, @sirupsen! The doc updates are incredibly helpful. 🎉

**Comment by hifi** [comment]
@sirupsen do you still have a test case you could throw at #416? I'm curious if you just hit an ordering bottleneck when retrieving WAL segments.

**Comment by sirupsen** [comment]
@hifi it's very fast for me these days despite the database being far larger. Probably with improper snapshot intervals and retention you might be able to make it slow!

**Comment by northeastprince** [comment]
Can this be closed?

**Comment by darkgnotic** [comment]
We're working with restores of very large (e.g. 100s of GB) SQLite databases, and in our case @erikmunson identified a bottleneck at the download of the large snapshot file:

https://github.com/benbjohnson/litestream/blob/996a50deaf0350e4450053bf54a8e350f37116c4/s3/replica_client.go#L285-L288

In AWS at least, there seems to be a bandwidth limit to a single connection, and s3 docs recommend breaking up downloads of large files into multiple segments:

https://docs.aws.amazon.com/code-library/latest/ug/s3_example_s3_Scenario_UsingLargeFiles_section.html

@benbjohnson if this makes sense, I can look into using the [Downloader](service/s3/s3manager/download.go) in the s3 SDK and repurposing the `parallelism` flag to download segments of snapshot file in parallel segments.

**Comment by darkgnotic** [comment]
Looks like this may not be that simple. `c.s3.GetObjectWithContext()` returns a `ReaderCloser` allowing litestream to stream the data from s3 and decompress it before writing it to disk. The `Downloader`, on the other hand, downloads the entire file before returning, requiring the caller to either provide `WriterAt` that can either be a destination file or an in-memory buffer. Neither is ideal as it would require either extra disk space or enough memory to hold the entire object.  😖

**Comment by darkgnotic** [comment]
I have multipart downloads this working [in a fork](https://github.com/benbjohnson/litestream/compare/main...darkgnotic:litestream:darkgnotic/multi-part-download) and confirmed that it's effective.

In AWS, a plain `litestream restore` of a 100GB db takes ~35 minutes.

With multipart downloads, i.e.

`litestream restore -multipart-concurrency 48 -multipart-size 16777216`

the time goes down to 5 minutes. 

Essentially this allows using an in-memory buffer of `concurrency * size` bytes when downloading / decompressing the snapshot, vs squeezing it through Golang's default 16kb socket buffer.

This is certainly more of a feature than a bug fix, so given the [Contribution Policy](https://github.com/benbjohnson/litestream?tab=readme-ov-file#contribution-policy) we'll be maintaining this in our own fork.

However, if you think this is something worth merging into the main code, I'd be happy to go through the PR process.

**Comment by benbjohnson** [comment]
This will improve in v0.5.0 as we have replaced the storage format & it now performs multi-level compactions.
